lib/vector-core/src/fanout.rs

```diff
@@ -63,165 +62,330 @@ impl Fanout {
         self.sinks.push((id, Some(sink)));
     }
 
+    /// Remove an existing sink as an output.
     fn remove(&mut self, id: &ComponentKey) {
-        let i = self.sinks.iter().position(|(n, _)| n == id);
-        let i = i.expect("Didn't find output in fanout");
-
-        let (_id, removed) = self.sinks.remove(i);
+        if let Some(i) = self.sinks.iter().position(|(n, _)| n == id) {
+            let (_id, _removed) = self.sinks.remove(i);
+        }
```

Comment:
This does change the contract some. Previously we'd kill the program, which indicated it was a serious, unrecoverable logic bug? I'm not opposed, more just pointing it out.

Reply:
Yeah, this should likely be tweaked. Initially I removed the panic because it was possible that the sink was present, but part of an in-flight send and just not in this particular place. That's still true some of the time, but not true when we're processing control messages before initiating a new send. I'll circle back and try to clean up some of the control message handling to make it more clear where we're asserting on invariants.

Reply:
Addressed in [27cbd3e](https://github.com/vectordotdev/vector/pull/11348/commits/27cbd3e5375711e706626c9a79379a3f2b5aa413)